lib/deflation.cpp

```diff
@@ -0,0 +1,527 @@
+        vi_global.push_back(accum);
+
+        for(int j = 0; j < local_length; j++) alpha[j] = -alpha[j]; 
+
+        blas::caxpy(alpha, vj_global, vi_global); //i-<j,i>j
+
+        offset += cdot_pipeline_length;
+      }
+
+      //alpha[0] = blas::norm2(param.RV->Component(i));
+      alpha[0] = blas::norm2(*accum);
+
+      //if(alpha[0].real() > 1e-16) blas::ax(1.0 /sqrt(alpha[0].real()), param.RV->Component(i));
+      if(alpha[0].real() > 1e-16) blas::ax(1.0 /sqrt(alpha[0].real()), *accum);
+      else                        errorQuda("\nCannot orthogonalize %dth vector\n", i);
+
+      param.RV->Component(i) = *accum;
+
+      param.matDeflation(*Av_sloppy, param.RV->Component(i));//precision must match!
+      //load diagonal:
+      //param.matProj[i*param.ld+i] = cDotProduct(param.RV->Component(i), *Av);
+      *Av = *Av_sloppy; 
+      param.matProj[i*param.ld+i] = cDotProduct(*accum, *Av);
+
+      //off-diagonal (use multiblas):
+      offset = 0;
+
+      while (offset < i){
+        const int local_length = (i - offset) > cdot_pipeline_length  ? cdot_pipeline_length : (i - offset) ;
+
+        std::vector<cudaColorSpinorField*> vj_local;
+        std::vector<cudaColorSpinorField*> av_local;
+
+        vj_local.reserve(local_length);
+        av_local.reserve(local_length);
+
+        for(int j = 0; j < local_length; j++)
+        {
+          vj_local.push_back(static_cast<cudaColorSpinorField*>(&param.RV->Component(offset+j)));
+          av_local.push_back(static_cast<cudaColorSpinorField*>(Av_sloppy));
+          //av_local.push_back(static_cast<cudaColorSpinorField*>(Av));
+          alpha[j] = 0.0;
+        }
+
+        //Warning! this won't work with arbitrary param.cur_dim. Pipelining is needed.
+        blas::cDotProduct(alpha, vj_local, av_local);
+
+        for (int j = 0; j < local_length; j++)//row id
+        {
+          param.matProj[i*param.ld+(j+offset)] = alpha[j];
+          param.matProj[(j+offset)*param.ld+i] = conj(alpha[j]);//conj
+        }
+
+        offset += cdot_pipeline_length;
+
+      }
+
+      delete [] alpha;
+    }
+
+    param.cur_dim += nev;
+
+    printfQuda("\nNew curr deflation space dim = %d\n", param.cur_dim);
+#endif
+    return;
+  }
+
+
+  void Deflation::reduce(double tol, int max_nev) {
+#ifdef DEFLATEDSOLVER
+     if(param.cur_dim < max_nev)
+     {
+        printf("\nToo big number of eigenvectors was requested, switched to maximum available number %d\n", param.cur_dim);
+        max_nev = param.cur_dim;
+     }
+
+     double *evals   = (double*)calloc(param.cur_dim, sizeof(double));//WARNING: Ritz values always in double.
+
+     Complex *projm  = (Complex*)mapped_malloc(param.ld*param.cur_dim * sizeof(Complex));
+     memcpy(projm, param.matProj, param.ld*param.cur_dim*sizeof(Complex));
+
+#ifdef MAGMA_LIB
+     //cudaHostRegister(static_cast<void *>(projm), param.ld*param.cur_dim*sizeof(Complex),  cudaHostRegisterDefault);
```

Comment:
Should this be deleted?  If it is left for future experiments, can you add a comment what it's for?